libs/recache/src/handler.rs

```diff
@@ -0,0 +1,216 @@
+use crate::stores::AsyncCacheStore;
+use bon::Builder;
+use dashmap::DashMap;
+use futures::{
+    FutureExt,
+    future::{BoxFuture, Shared},
+};
+use std::{cmp::Eq, fmt::Display, hash::Hash, sync::Arc, time::Duration};
+use thiserror::Error;
+
+type SharedFuture<V, E> = Shared<BoxFuture<'static, Result<V, E>>>;
+type EventHandler<V> = Arc<dyn Fn(&V) + Send + Sync>;
+type InflightRequest<K, V, C> =
+    SharedFuture<V, CacheRequestError<<C as AsyncCacheStore<K, V>>::Error>>;
+type InflightMap<K, V, C> = DashMap<K, InflightRequest<K, V, C>>;
+
+#[derive(Builder)]
+pub struct CacheHandler<C, K, V>
+where
+    C: AsyncCacheStore<K, V>,
+    K: Sync + Hash + Eq,
+{
+    #[builder(start_fn)]
+    cache: Arc<C>,
+    #[builder(default)]
+    inflight: Arc<InflightMap<K, V, C>>,
+    on_hit: Option<EventHandler<V>>,
+    on_computed: Option<EventHandler<V>>,
+    on_refresh_computed: Option<EventHandler<V>>,
```

Comment:
Nit: what do you think about having some no-op implementation set by default instead of `Option`s? That allows to get rid of `if let Some(handler) = on_handler` lines and just call the function as default no-op implementation will be used in None case

Reply:
That looks nice, I guess the compiler can optimize such noop function calls anyway